### ISSUE TITLE
Customizable Variant of PackedDoubleCoordinateSequence

### ISSUE THREAD

**Issue by airbreather** [NetTopologySuite/NetTopologySuite#482]
Discussing with @bjornharrtell on Gitter, it seems like there's a need for a generalized variant of `DotSpatialAffineCoordinateSequence` that supports (minimally) `ArraySegment<double>` rather than forcing the data to exactly cover an entire array.

In my own research, I've also found a use for a `CoordinateSequence` implementation that supports a memory layout where X and Y are in their own parallel arrays (particularly for pointwise SIMD operations).

Even `PackedDoubleCoordinateSequence`-style layouts would benefit from being able to be created from slices of existing arrays instead of always having to fully populate the array.

To that end, I'm going to get started on a coordinate sequence (+ factory) that supports more-or-less any densely packed layout backed by `Memory<double>` instances.  The requirements will be, roughly:
1. For any given dimension, the entire contents must fit within a single `Memory<double>`.
1. No gaps or scrambling: `array[0]` must hold the first value in one dimension, `array[1]` must hold either the first value in another dimension or (if the array holds data for only a single dimension) the second value in the first dimension, and so on.

**Comment by airbreather** [comment]
Started a branch: https://github.com/NetTopologySuite/NetTopologySuite/tree/gh-482

It's mostly working, just need more in-depth testing I think.

**Comment by bjornharrtell** [comment]
Looks good on first glance. I will take it for a spin and get back with my findings.

**Comment by bjornharrtell** [comment]
When used in`IO.PostGIS` I get the following benchmark results from naive usage (also without the optimizations I did recently for PackedCoordinateSequenceFactory:

|                                   Method |     Mean |     Error |    StdDev |
|----------------------------------------- |---------:|----------:|----------:|
|                         RoundtripDefault | 2.612 us | 0.0271 us | 0.0241 us |
| RoundtripPackedCoordinateSequenceFactory | 2.273 us | 0.0238 us | 0.0211 us |
|    RoundtripRawCoordinateSequenceFactory | 3.664 us | 0.0388 us | 0.0324 us |

So the the abstraction unfortunately seem to come with a fair bit of overhead.

I haven't been able to use it in an optimal way yet as it is not obvious to me how to use it with existing raw data.

**Comment by bjornharrtell** [comment]
Unless I misunderstand your design does not allow for the same memory layout as in PackedCoordinateSequence?

**Comment by bjornharrtell** [comment]
I figured out how to use it more or less directly with existing raw data but they way I use it is hardcoded to the benchmark data so it's a bit of a hack. Still a bit of non-zero-copy when reading that could possibly be removed/reduced by a custom MemoryManager.

Results are nicer but I'm not convinced it is a good idea at all to abstract/generalize in this way the naive usage is worse than the default and the usage of the sequence becomes complex.

|                                   Method |     Mean |     Error |    StdDev |
|----------------------------------------- |---------:|----------:|----------:|
|                         RoundtripDefault | 2.666 us | 0.0375 us | 0.0332 us |
| RoundtripPackedCoordinateSequenceFactory | 1.099 us | 0.0177 us | 0.0148 us |
|    RoundtripRawCoordinateSequenceFactory | 1.542 us | 0.0195 us | 0.0182 us |

**Comment by airbreather** [comment]
> Results are nicer (also includes the optimized path for `PackedCoordinateSequence`) but I'm not convinced it is a good idea at all to abstract/generalize in this way the naive usage is worse than the default and the usage of the sequence becomes complex.

The sequence has a lot more fixed-cost overhead than other sequences.  The absolute difference between the fastest and slowest in your benchmark is under 2 microseconds.  Add more data (and `[MemoryDiagnoser]`), switch to .NET 5.0 for the benchmark, and you'll start to see why I bothered.

The code from NetTopologySuite/NetTopologySuite.IO.PostGis@d26d9af gives me:

``` ini

BenchmarkDotNet=v0.12.1, OS=linuxmint 20.1
Intel Core i7-6850K CPU 3.60GHz (Skylake), 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=5.0.103
  [Host]     : .NET Core 5.0.3 (CoreCLR 5.0.321.7203, CoreFX 5.0.321.7203), X64 RyuJIT
  Job-ATQSZF : .NET Core 5.0.3 (CoreCLR 5.0.321.7203, CoreFX 5.0.321.7203), X64 RyuJIT

Arguments=/p:GenerateProgramFile=false  

```
|                                   Method |      Mean |     Error |    StdDev |  Gen 0 |  Gen 1 | Gen 2 | Allocated |
|----------------------------------------- |----------:|----------:|----------:|-------:|-------:|------:|----------:|
|                         RoundtripDefault | 44.757 μs | 0.3134 μs | 0.2447 μs | 5.1270 | 0.5493 |     - |   40288 B |
| RoundtripPackedCoordinateSequenceFactory |  3.584 μs | 0.0651 μs | 0.0609 μs | 5.1651 | 0.2937 |     - |   40512 B |
|    RoundtripRawCoordinateSequenceFactory |  2.757 μs | 0.0172 μs | 0.0153 μs | 3.2120 | 0.1526 |     - |   25208 B |

**Comment by bjornharrtell** [comment]
Interesting! But does that mean that it will remain suboptimal for large volumes of individual points?

**Comment by airbreather** [comment]
> does that mean that it will remain suboptimal for large volumes of individual points?

Yes.  Large volumes of individual points perform poorly in NTS in general, and they would be handled particularly poorly by a sequence such as this.

> don't you agree it's to difficult for end users to use in an optimal way

Yes and no.  End users will tend to call higher-order methods such as `PostGisReader.Read`, rather than implementing them themselves, so reading / writing are non-issues.

However, it's true that it is difficult to get the full benefit of this kind of coordinate sequence, as you pretty much have to go from the raw coordinates w/ stride in order to minimize the `Memory<double>` to `Span<double>` conversions and related pointer chasing that eat up so much of the execution time.  This is not a common pattern, especially in .NET.

That said, the **potential** benefits are significant:
1. `Memory<double>` can be backed by a memory-mapped file in order to "read" a coordinate sequence without a commit charge or immediately touching the underlying data.  The operating system can page in just the data for the geometries that you use.  Of course, this would lead to unpredictable execution times, and I think I/O errors turn into access violations which are really spooky, but I'm sure there are use cases.
1. For a layout where **all** dimensions are stored separately, pointwise operations could potentially benefit from SIMD instructions.  I have some benchmarks [over here](https://github.com/airbreather/AirBreather.Projections) from a few years ago proving that AVX2 cut off 1/3 of the runtime of a particular operation, compared to plain loop unrolling.

**Comment by bjornharrtell** [comment]
Ok @airbreather you got me convinced it might be worth the tradeoffs, I'm experimenting with it now and gathering more benchmarks etc.

**Comment by bjornharrtell** [comment]
I've done some more experimentation. As my goal is to find an optimal CoordinateSequence for use in the FlatGeobuf C# reference implementation I've benchmarked in that context. I'm still not sold on the performance of it with DotSpatial winning with quite some margin. Here is a recent run of a benchmark parameterized on number of vertices:

BenchmarkDotNet=v0.12.1, OS=ubuntu 20.10
Intel Core i7-1065G7 CPU 1.30GHz, 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=5.0.103
  [Host]     : .NET Core 5.0.3 (CoreCLR 5.0.321.7203, CoreFX 5.0.321.7203), X64 RyuJIT
  DefaultJob : .NET Core 5.0.3 (CoreCLR 5.0.321.7203, CoreFX 5.0.321.7203), X64 RyuJIT

|                Method | Vertices |   Sequence |           Mean |        Error |       StdDev |
|---------------------- |--------- |----------- |---------------:|-------------:|-------------:|
| LineStringDeserialize |        2 |    Default |       353.0 ns |      3.14 ns |      2.78 ns |
| LineStringDeserialize |        2 | DotSpatial |       239.7 ns |      1.77 ns |      1.57 ns |
| LineStringDeserialize |        2 |        Raw |       288.4 ns |      2.87 ns |      2.68 ns |
| LineStringDeserialize |       20 |    Default |       869.8 ns |      4.66 ns |      3.89 ns |
| LineStringDeserialize |       20 | DotSpatial |       313.8 ns |      2.32 ns |      1.81 ns |
| LineStringDeserialize |       20 |        Raw |       619.7 ns |      6.53 ns |      5.79 ns |
| LineStringDeserialize |      200 |    Default |     5,821.4 ns |     35.87 ns |     31.80 ns |
| LineStringDeserialize |      200 | DotSpatial |       998.8 ns |      5.32 ns |      4.97 ns |
| LineStringDeserialize |      200 |        Raw |     3,767.7 ns |     11.35 ns |      9.48 ns |
| LineStringDeserialize |    20000 |    Default | 1,160,276.3 ns | 19,964.95 ns | 18,675.23 ns |
| LineStringDeserialize |    20000 | DotSpatial |   140,073.9 ns |  2,098.98 ns |  1,963.39 ns |
| LineStringDeserialize |    20000 |        Raw |   348,273.7 ns |  1,649.55 ns |  1,462.28 ns |

**Comment by bjornharrtell** [comment]
@airbreather not my intention to be negative, I find your implementation impressive and it has improved my knowledge of Span/Memory and I'm grateful for that. But it seems plain arrays is the faster alternative for now.

**Comment by airbreather** [comment]
> @airbreather not my intention to be negative, I find your implementation impressive and it has improved my knowledge of Span/Memory and I'm grateful for that. But it seems plain arrays is the faster alternative for now.

No worries -- it's not surprising to find that a component that's restricted to (and optimized for) a particular use case / memory layout will perform better in its niche than a more flexible component.  It's also, of course, very useful to have numbers to talk about, as opposed to guesses.

I'm more interested in knowing about use cases where it's unacceptably slow... looking at this table, even the worst performer on the largest data set loads in just barely over 1ms.  In the context of what we're talking about here:

1. The data has to come from **somewhere**.  If it's coming from over the network (or even a cold file on a filesystem), the I/O overhead is going to turn all of these differences into noise.
1. In theory, you're loading the geometry in order to actually **do something** with it... most of the interesting "somethings" are likely to dwarf that I/O overhead.
1. Even if I'm wrong, and the deserialization CPU time represents a significant fraction of the time it takes to run some operation, I can't help but wonder how many operations need to be running in parallel on a server before the 1ms difference between the best and the worst performer adds up to serious losses...
1. In situations where the flexibility of `Memory<double>` would allow us to write it in a way that reduces pressure on the garbage collector (i.e., that fancy `MemoryManager<T>` I wrote), these numbers can be somewhat misleading, because the cost of that added pressure is felt most strongly when the system is under heavy load.
    - This is why I added `[MemoryDiagnoser]` to my benchmark run: the "Allocated" column is an **INDICATOR** of the additional impact that this code would have on the rest of the system when it experiences heavy memory pressure.

So in short, I think that it would still be useful for us to make a `CoordinateSequence` implementation that's more flexible than the existing ones, even if it's not technically the fastest.

**Comment by bjornharrtell** [comment]
I agree with your observations and my (current) case might be a bit of a corner case. I've been writing an experimental lightweight WMS server in C# and the measurements I've been looking at is when data is from local SSD. I'm rendering with SkiaSharp which is very fast and so the deserialization does become significant.

**Comment by FObermaier** [comment]
I like that `RawCoordinateSequence` is performant while flexible.
I don't like that it seems to be difficult to unleash its potential.
I propose to add some `Create` functions to `RawCoordinateSequenceFactory` that faciliate the creation of `RawCoordinateSequence`s from a byte array for common memory layout:

```C#
    /// <summary>
    /// An enumeration of common data alignments for <see cref="RawCoordinateSequence"/>.
    /// </summary>
    public enum RawCoordinateSequenceDataAlignment
    {
        /// <summary>
        /// Data is stored as a sequence of intermittent x- and y-ordinate values
        /// </summary>
        XY,
        /// <summary>
        /// Data is stored as a sequence of intermittent longitude and latitude values
        /// </summary>
        LonLat = XY,
        /// <summary>
        /// Data is stored as a sequence of intermittent latitude and longitude values
        /// </summary>
        LatLon,
        /// <summary>
        /// Data is stored as a sequence of x-ordinate values followed by a sequence of y-ordinate values
        /// </summary>
        X_Y,
        /// <summary>
        /// Data is stored as a sequence of longitude values followed by a sequence of latitude values
        /// </summary>
        Lon_Lat = X_Y,
        /// <summary>
        /// Data is stored as a sequence of latitude values followed by a sequence of longitude values
        /// </summary>
        Lat_Lon,
        /// <summary>
        /// Data is stored as a sequence of intermittent x-, y- and z-ordinate values
        /// </summary>
        XYZ,
        /// <summary>
        /// Data is stored as a sequence of intermittent x-, y- and m-ordinate values
        /// </summary>
        XYM,
        /// <summary>
        /// Data is stored as a sequence of x-ordinate values followed by a sequence of y-ordinate values and a sequence of z-ordinate values
        /// </summary>
        X_Y_Z,
        /// <summary>
        /// Data is stored as a sequence of intermittent x-ordinate and y-ordinate values followed by a sequence of z-ordinate values
        /// </summary>
        XY_Z,
        /// <summary>
        /// Data is stored as a sequence of x-ordinate values followed by a sequence of y-ordinate values and a sequence of m-ordinate values
        /// </summary>
        X_Y_M,
        /// <summary>
        /// Data is stored as a sequence of intermittent x-ordinate and y-ordinate values followed by a sequence of m-ordinate values
        /// </summary>
        XY_M,
        /// <summary>
        /// Data is stored as a sequence of intermittent x-, y-, z- and m-ordinate values
        /// </summary>
        XYZM,
        /// <summary>
        /// Data is stored as a sequence of intermittent x- and y-ordinate values followed by a sequence of z-ordinate values and a sequence of m-ordinate values
        /// </summary>
        XY_Z_M,
        /// <summary>
        /// Data is stored as a sequence of x-ordinate values followed by a sequence of y-ordinate values, a sequence of z-ordinate values a sequence of m-ordinate values
        /// </summary>
        X_Y_Z_M,
    }
```
```C#
   public class RawCoordinateSequenceFactory : CoordinateSequenceFactory 
   {
        /*
         * current implementation
         */

        /// <summary>
        /// Creates a coordinate sequence for the given <paramref name="data"/> shaping to the given <paramref name="alignment"/>
        /// </summary>
        /// <param name="alignment">The alignment of ordinate data</param>
        /// <param name="data">The data in <see cref="Memory{T}"/></param>
        /// <returns>A coordinate sequence</returns>
        public CoordinateSequence Create(RawCoordinateSequenceDataAlignment alignment, Memory<double> data)
        {
            Prepare(alignment, data, out var dimensionMap, out var dataBlocks, out int measures);

            return new RawCoordinateSequence(dataBlocks, dimensionMap, measures);
        }

        /// <summary>
        /// Creates a coordinate sequence for the given <paramref name="data"/> shaping to the given <paramref name="alignment"/>
        /// </summary>
        /// <param name="alignment">The alignment of ordinate data</param>
        /// <param name="data">The data</param>
        /// <returns>A coordinate sequence</returns>
        public CoordinateSequence Create(RawCoordinateSequenceDataAlignment alignment, double[] data)
        {
            return Create(alignment, new Memory<double>(data));
        }

        /// <summary>
        /// Creates a coordinate sequence for the given <paramref name="data"/> shaping to the given <paramref name="alignment"/>
        /// </summary>
        /// <param name="alignment">The alignment of ordinate data</param>
        /// <param name="data">The data</param>
        /// <returns>A coordinate sequence</returns>
        public CoordinateSequence Create(RawCoordinateSequenceDataAlignment alignment, byte[] data)
        {
            // to Memory<double>
            var memory = new CastingMemoryManager<double>(new ArraySegment<byte>(data));

            return Create(alignment, memory.Memory);
        }

        private void Prepare(RawCoordinateSequenceDataAlignment alignment, Memory<double> data,
            out (int, int)[] dimensionMap, out Memory<double>[] dataBlocks, out int measures)
        {
            measures = 0;
            int size = data.Length;
            switch (alignment)
            {
                case RawCoordinateSequenceDataAlignment.XY:
                //case RawCoordinateSequenceDataAlignment.LonLat:
                    if (size % 2 != 0)
                        throw new ArgumentException($"Alignment {alignment} cannot match given {nameof(data)} (size={size}bytes)");
                    dimensionMap = new[] {(0, 0), (0, 1)};
                    dataBlocks = new[] {data};
                    return;
                case RawCoordinateSequenceDataAlignment.LatLon:
                    //case RawCoordinateSequenceDataAlignment.LonLat:
                    dimensionMap = new[] { (0, 1), (0, 0) };
                    dataBlocks = new[] { data };
                    return;
                case RawCoordinateSequenceDataAlignment.X_Y:
                    size /= 2;
                    dimensionMap = new[] { (0, 0), (1, 0) };
                    dataBlocks = new[] { data.Slice(0, size), data.Slice(size) };
                    return;
                case RawCoordinateSequenceDataAlignment.Lat_Lon:
                    //case RawCoordinateSequenceDataAlignment.LonLat:
                    dimensionMap = new[] { (1, 0), (0, 0) };
                    dataBlocks = new[] { data };
                    return;

                // To Be continued...
            }

            throw new NotImplementedException();
        }
   }
```

**Comment by FObermaier** [comment]
On my way to work, i thought a more general enum would be better:
```C#
public enum RawCoordinateSequenceAlignment
{
    /// <summary>
    /// All ordinate values are kept in seperate contiguous data blocks
    /// </summary>
   Seperate,
    /// <summary>
    /// All ordinate values are kept intermittent in one contiguous data block
    /// </summary>
   Intermittent,
    /// <summary>
    /// X- and y-ordinate values are kept intermittent in one contiguous data block, all
    /// other ordinate values have their own data block
    /// </summary>
   IntermittentXY,
}
```
That could serve as property `RawCoordinateSequenceFactory` and be used in the `Create(int size, int dimension, int measures)` overload.

**Comment by FObermaier** [comment]
Sorry for the noise, I seem to not have grasped the whole of `RawCoordinateSequence` and its factory class.
Now, i still think that `RawCoordinateSequenceFactory` lacks a `Create(int size, int dimension, int measures, double[] ordinateData)` overload. Together with `RawCoordinateSequenceFactory.OrdinateGroups` the `ordinateData` could be transformed to a `RawCoordinateSequence` internally.

**Comment by airbreather** [comment]
`RawCoordinateSequenceFactory.OrdinateGroups` indicates how the data is grouped together in the **output** sequences.  IMO, something like `Create(int size, int dimension, int measures, double[] ordinateData)` would be too easy to use incorrectly, because it makes an implicit demand on the caller to pass in data that conforms to the layout that's implied by `OrdinateGroups`.

I do think that it would be great to have comparatively easy ways to create these sequences.  Perhaps something like:
```csharp
public RawCoordinateSequence CreateXY(Memory<double> x, Memory<double> y)
{
    /* ... */
}

public RawCoordinateSequence CreateXY(Memory<double> xy)
{
    /* ... */
}

public RawCoordinateSequence CreateXYZ(Memory<double> x, Memory<double> y, Memory<double> z)
{
    /* ... */
}

public RawCoordinateSequence CreateXYZ(Memory<double> xy, Memory<double> z)
{
    /* ... */
}

public RawCoordinateSequence CreateXYZ(Memory<double> xyz)
{
    /* ... */
}

public RawCoordinateSequence CreateXYM(Memory<double> x, Memory<double> y, Memory<double> m)
{
    /* ... */
}

public RawCoordinateSequence CreateXYM(Memory<double> xy, Memory<double> m)
{
    /* ... */
}

public RawCoordinateSequence CreateXYM(Memory<double> xym)
{
    /* ... */
}

public RawCoordinateSequence CreateXYZM(Memory<double> x, Memory<double> y, Memory<double> z, Memory<double> m)
{
    /* ... */
}

public RawCoordinateSequence CreateXYZM(Memory<double> xy, Memory<double> z, Memory<double> m)
{
    /* ... */
}

public RawCoordinateSequence CreateXYZM(Memory<double> xyzm)
{
    /* ... */
}
```
